nilearn/decoding/decoder.py

```diff
@@ -0,0 +1,818 @@
+from sklearn.linear_model.ridge import Ridge, RidgeClassifier, _BaseRidge
+from sklearn.linear_model import LogisticRegression
+from sklearn.svm import LinearSVC, SVR
+from sklearn.preprocessing import LabelBinarizer
+from sklearn.svm.bounds import l1_min_c
+from sklearn.base import RegressorMixin
+from sklearn.linear_model.base import LinearModel
+from sklearn.utils.extmath import safe_sparse_dot
+from sklearn import clone
+from sklearn.model_selection import ParameterGrid
+from sklearn.utils.validation import check_is_fitted, check_X_y
+from sklearn.metrics import check_scoring
+from sklearn.model_selection import check_cv
+
+from ..input_data.masker_validation import check_embedded_nifti_masker
+from .._utils.param_validation import _adjust_screening_percentile
+from .._utils.compat import _basestring
+from .._utils.cache_mixin import _check_memory
+from .._utils.param_validation import check_feature_screening
+from .._utils import CacheMixin
+
+
+SUPPORTED_ESTIMATORS = dict(
+    svc_l1=LinearSVC(penalty='l1', dual=False),
+    svc_l2=LinearSVC(penalty='l2'),
+    svc=LinearSVC(penalty='l2'),
+    logistic_l1=LogisticRegression(penalty='l1', solver='liblinear'),
+    logistic_l2=LogisticRegression(penalty='l2', solver='liblinear'),
+    logistic=LogisticRegression(penalty='l2'),
+    ridge_classifier=RidgeClassifier(),
+    ridge_regression=Ridge(),
+    ridge=Ridge(),
+    svr=SVR(kernel='linear'),
+)
+
+
+def _check_param_grid(estimator, X, y, param_grid):
+    """Check param_grid and return sensible default if none is given.
+    For more information, see the documentation:
+    <http://http://scikit-learn.org/stable/modules/generated/sklearn.svm.l1_min_c>
+    """
+    if param_grid is None:
+        param_grid = {}
+        # define loss function
+        if isinstance(estimator, LogisticRegression):
+            loss = 'log'
+        elif isinstance(estimator, (LinearSVC, _BaseRidge, SVR)):
+            loss = 'squared_hinge'
+        else:
+            raise ValueError("%s is an unknown estimator."
+                             "The supported estimators are: %s" %
+                             (estimator, list(SUPPORTED_ESTIMATORS.keys())))
+
+        # loss = l2 was deprecated after version 0.17
+        if LooseVersion(sklearn.__version__) <= LooseVersion('0.17'):
+            if loss == 'squared_hinge':
+                loss = 'l2'
+
+        if hasattr(estimator, 'penalty') and (estimator.penalty == 'l1'):
+            min_c = l1_min_c(X, y, loss=loss)
+        else:
+            min_c = 0.5
+
+        if not isinstance(estimator, _BaseRidge):
+            param_grid['C'] = np.array([2, 20, 200]) * min_c
+        else:
+            param_grid['alpha'] = 1. / np.array([1, 10, 100])
+
+    return param_grid
+
+
+def _parallel_fit(estimator, X, y, train, test, param_grid, is_classif, scorer,
+                  mask_img, class_index, screening_percentile=None):
+    """Find the best estimator for a fold within a job.
```

Comment:
The docstring is hard to get because it mixes the role of the function for the object and what it does exactly.

I think it'd be good to state clearly that :
This function tries several parameters for the estimator for the train and test fold provided and save the one that performs best.